lib/db/mysql.js

```diff
@@ -413,92 +407,148 @@ module.exports = function (log, error) {
     })
   }
 
-  // This function takes a result and unwraps comma-separated values to Arrays
-  // and eventually applies transforms.
-  function unwrapArrays(result, fields, separator = ',') {
-    for (const {name, transform} of fields) {
-      const arr = result[name] ? result[name].split(separator) : []
-      result[name] = transform ? transform(arr) : arr
-    }
-  }
-
-  function unwrapCapabilities(result, fieldName) {
-    const field = {
-      name: fieldName,
-      transform(capabilityIds) {
-        return capabilityIds.reduce((acc, c) => {
-          const capabilityStr = dbUtil.mapDeviceCapability(parseInt(c))
-          if (capabilityStr != null) {
-            acc.push(capabilityStr)
+  // A helper function for aggregating name:value pairs into a JSON object.
+  // It takes a list of rows like this:
+  //
+  //   +-------+-------+-------+--------+
+  //   | id    | other | name  | value  |
+  //   +-------+-------+-------+--------+
+  //   | one   | foo   | name1 | value1 |
+  //   | two   | bar   | name1 | value1 |
+  //   | two   | bar   | name2 | value2 |
+  //   | three | baz   | NULL  | NULL   |
+  //   +-------+-------+-------+--------+
+  //
+  // And aggregates the name/data pairs into a JSON object like this:
+  //
+  //  [
+  //    { id: 'one', other: 'foo', agg: { 'name1': 'value1' }},
+  //    { id: 'two', other: 'bar', agg: { 'name1': 'value1', 'name2': 'value1' }},
+  //    { id: 'three', other: 'baz', agg: {}},
+  //  ]
+  //
+  // Note that correct operation depends on the rows being ordered by id column,
+  // so they can be aggregated in a single pass.
+  //
+  // The column names can be customized, in order to work with several different
+  // queries.
```

Comment:
Even with this and the inline commentary, I'm finding it hard to follow what this function does. Few suggestions, feel free to ignore:

* If this comment was more docblocky, with definitions for each argument etc, I think I'd find it easier to understand. For instance, what is `aggColumn`? From reading the code, I think it's just the name of the property that is mapped to `agg` in your example output above, but it would be good to have those definitions laid out formally.

* Alternatively, just add the example function call that produces the output above. We have the input definition and the output definition, so maybe just seeing what the function call is with hard-coded values for the `...Column` params might be enough.

* It might be easier to follow if the example used real-life column names, keys and values from the db schema. The fact that the example data is so abstract makes it hard to pin down in my head.

* It's poor reading comprehension on my part, but it took me ages to spot that `id` is not unique in the example input data (only just spotted it while I was typing the last bullet). That's the key transformation here right, we're normalizing the data so that each `id` is unique in the result set? Might be worth calling that out more explicitly, if I understand it correctly.

Reply:
(and if the last point is on the right track, I wonder whether calling the function `normalizeNameValuePairs` might help too? or is the word "normalize" just imbued with unwarranted significance in my head?)